### DISCUSSION TITLE
Information to restore deleted branches

### DISCUSSION THREAD

**Discussion by ilyagr** [magit/magit#4339] [Ideas]
I've gotten used to Magit holding my hand and it being easy to reverse anything. So, I was surprised to find it is in fact possible to mess up in a way that's difficult to undo (though no permanent damage was done). I wonder if Magit could hold my hand some more :)

I am using Magit 20210209.1110. I had fetched from the remote, and had an `origin/blah` branch in my log that I didn't care about. So, I erased it with `b x`, wrongly imagining that the branch will merely be forgotten locally. Instead, Magit pushed the deletion to the remote immediately.

This is difficult to undo. There is no reflog because the branch is gone. `git` should print a message telling you what commit the deleted branch was (or does it? This was a `git push` command, not `git branch`), but by the time I realized that, the whole interaction was gone from `magit-process-buffer`.  I use `magit-wip-mode`, but it was no help since I never checked out that branch. `git fsck` may have been an option, but I was not that desperate (I had another machine that still had this branch). 

Obviously, this is not Magit's fault, but I started wondering what Magit could do differently. All of these things are in the "nice to have" category, and each one could be useful on its own.

1. Try to solve this with logging. Some options:
  
    a. Recorded  the commit hash of the deleted branch in a message shown to the user  (both for `git branch` deletes and `git push` deletes). It would then also go to Emacs's messages buffer. Currently, it just says `Git finished`.

    b. Make the `magit-process-buffer` log history longer, and perhaps more persistent. If `git push` does not report the commit hash on remote branch deletion, Magit could put it there instead.

    c. Have a special kind of log just for the purpose of recovering deleted branches (local or remote). Have this log persist when Emacs is restarted.

2. I use `magit-wip-mode`, but it was no help. It could mark deleted branches with its special kind of ref.

3. Perhaps have a nice way to do what I was trying to do originally (get rid of the `origin/blah` branch in my local copy) and/or a more severe warning when you're deleting a branch on the remote.

**Comment by tarsius** [comment]
:grin: `magit-branch-delete` can always be improved...

```elisp
(defun magit-branch-delete (branches &optional force)
  "Delete one or multiple branches.
If the region marks multiple branches, then offer to delete
those, otherwise prompt for a single branch to be deleted,
defaulting to the branch at point."
  ;; One would expect this to be a command as simple as, for example,
  ;; `magit-branch-rename'; but it turns out everyone wants to squeeze
  ;; a bit of extra functionality into this one, including myself.
  ...92 lines...)
```

**Reply by tarsius** [reply]
I went with 3 and some variation of 1 (the hash-note goes into the `*Messages*`) buffer.

**Reply by ilyagr** [reply]
That is great, thank you very much!

Mainly for my reference, it's https://github.com/magit/magit/commit/07d9c2c47639e855e00b63cdce2513b67e488516 and https://github.com/magit/magit/commit/d14c082354c796dd363550a03463e9382b7cfbf6.

I may, at some point, try to do 2) as well, though probably not immediately. I'd need to find some time and then to figure out how one gets emacs to run a checked out version of magit.
